Unable to remove program from list

Clicking the remove button causes and error.

empty programs can be removed without problems.
confirmed for programs with program years, this errors out with the following message

`error.js:28 Error: Assertion Failed: Cannot call get with 'length' on an undefined object.`

--- a/app/models/program.js
+++ b/app/models/program.js
@@ -3,7 +3,6 @@ import DS from 'ember-data';
 import PublishableModel from 'ilios/mixins/publishable-model';
 
 const { computed } = Ember;
-const { mapBy, sum } = computed;
 
 export default DS.Model.extend(PublishableModel,{
   title: DS.attr('string'),
@@ -15,10 +14,68 @@ export default DS.Model.extend(PublishableModel,{
     inverse: 'program'
   }),
   curriculumInventoryReports: DS.hasMany('curriculum-inventory-report', {async: true}),
-  cohorts: mapBy('programYears', 'cohort'),
-  courses: mapBy('cohorts', 'courses'),
-  courseCounts: mapBy('courses', 'length'),
-  courseCount: sum('courseCounts'),
+
+  cohorts: computed('programYears.[]', function() {
+    let defer = Ember.RSVP.defer();
+    let allCohorts = [];
+    let promises = [];
+    promises.pushObject(new Ember.RSVP.Promise(resolve => {
+      this.get('programYears').then(programYears => {
+        if(!programYears.length){
+          resolve();
+        }
+        let promises = [];
+        programYears.forEach(programYear => {
+          promises.pushObject(programYear.get('cohort').then(cohort =>{
+            allCohorts.pushObject(cohort);
+          }));
+        });
+        Ember.RSVP.all(promises).then(()=>{
+          resolve();
+        });
+      });
+    }));
+
+    Ember.RSVP.all(promises).then(()=>{
+      defer.resolve(allCohorts);
+    });
+
+    return DS.PromiseArray.create({
+      promise: defer.promise
+    });
+  }),
+
+  courses: computed('cohorts.[]', function() {
+    let defer = Ember.RSVP.defer();
+    let allCourses = [];
+    let promises = [];
+    promises.pushObject(new Ember.RSVP.Promise(resolve => {
+      this.get('cohorts').then(cohorts => {
+        if(!cohorts.length){
+          resolve();
+        }
+        let promises = [];
+        cohorts.forEach(cohort => {
+          promises.pushObject(cohort.get('courses').then(courses =>{
+            courses.forEach(course => {
+              allCourses.pushObject(course);
+            });
+          }));
+        });
+        Ember.RSVP.all(promises).then(()=>{
+          resolve();
+        });
+      });
+    }));
+
+    Ember.RSVP.all(promises).then(()=>{
+      defer.resolve(allCourses.uniq());
+    });
+
+    return DS.PromiseArray.create({
+      promise: defer.promise
+    });
+  }),
   requiredPublicationSetFields: ['title', 'shortTitle', 'duration'],
   optionalPublicationLengthFields: ['programYears'],
   requiredPublicationIssues: computed('title', 'shortTitle', 'duration', function(){

--- a/app/templates/components/program-list.hbs
+++ b/app/templates/components/program-list.hbs
@@ -25,15 +25,21 @@
         </td>
       </tr>
       {{#if program.showRemoveConfirmation}}
-        <tr class='confirm-removal'><td colspan=9>
-          <div class='confirm-message'>
-            {{t 'programs.confirmRemove' programYearCount=program.programYears.length courseCount=program.courseCount}} <br />
-           <div class="confirm-buttons">
-            <button {{action 'remove' program}} class='remove text'>{{t 'general.yes'}}</button>
-            <button {{action 'cancelRemove' program}} class='done text'>{{t 'general.cancel'}}</button>
-          </div>
-          </div>
-        </td></tr>
+        <tr class='confirm-removal'>
+          <td colspan=9>
+            <div class='confirm-message'>
+            {{#if (is-fulfilled program.courses)}}
+              {{t 'programs.confirmRemove' programYearCount=program.programYears.length courseCount=program.courses.length}} <br />
+              <div class="confirm-buttons">
+                <button {{action 'remove' program}} class='remove text'>{{t 'general.yes'}}</button>
+                <button {{action 'cancelRemove' program}} class='done text'>{{t 'general.cancel'}}</button>
+              </div>
+            {{else}}
+              <div>{{fa-icon 'spinner' spin=true}}</div>
+            {{/if}}
+            </div>
+          </td>
+        </tr>
       {{/if}}
     {{/each}}
   </tbody>

